anndata/readwrite/utils.py

```diff
@@ -196,9 +196,6 @@ def func_wrapper(elem, key, val, *args, **kwargs):
                 raise type(e)("\n\n".join((
                     str(e),
                     f"Above error raised while writing key '{key}' of type {type(val)} from {parent}."
-                )))
-                # raise AnnDataReadError(
-                #     f"Above error raised while writing key '{key}' of type {type(val)} from {parent}."
-                # )
+                ))).with_traceback(sys.exc_info()[2])
```

Comment:
Unclear what this is supposed to do. “Thrown twice”? If you mean you don’t want to see “this exception was caused by”, you should use [PEP 3134](https://www.python.org/dev/peps/pep-3134/#explicit-exception-chaining): `raise ... from e` or `raise ... from e.__cause__` or so.

Reply:
<table>
<tr>
<td>

```pytb
In [12]: try: 
    ...:     try: 
    ...:         raise Exception('test') 
    ...:     except Exception: 
    ...:         raise Exception('test2') 
    ...: except Exception as e: 
    ...:     raise Exception(f'{e} 3') from e 
    ...:      
    ...:                                                                                                                                                                                                                                     
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-12-6e2b8f9e70be> in <module>
      2     try:
----> 3         raise Exception('test')
      4     except Exception:

Exception: test

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
<ipython-input-12-6e2b8f9e70be> in <module>
      4     except Exception:
----> 5         raise Exception('test2')
      6 except Exception as e:

Exception: test2

The above exception was the direct cause of the following exception:

Exception                                 Traceback (most recent call last)
<ipython-input-12-6e2b8f9e70be> in <module>
      5         raise Exception('test2')
      6 except Exception as e:
----> 7     raise Exception(f'{e} 3') from e
      8 
      9 

Exception: test2 3
```

</td><td>

```pytb
In [13]: try: 
    ...:     try: 
    ...:         raise Exception('test') 
    ...:     except Exception: 
    ...:         raise Exception('test2') 
    ...: except Exception as e: 
    ...:     raise Exception(f'{e} 3').with_traceback(sys.exc_info()[2]) 
    ...:                                                                                                                                                                                                                                     
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-13-8dde79ff88d6> in <module>
      2     try:
----> 3         raise Exception('test')
      4     except Exception:

Exception: test

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
<ipython-input-13-8dde79ff88d6> in <module>
      4     except Exception:
----> 5         raise Exception('test2')
      6 except Exception as e:

Exception: test2

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
<ipython-input-13-8dde79ff88d6> in <module>
      5         raise Exception('test2')
      6 except Exception as e:
----> 7     raise Exception(f'{e} 3').with_traceback(sys.exc_info()[2])
      8 

<ipython-input-13-8dde79ff88d6> in <module>
      3         raise Exception('test')
      4     except Exception:
----> 5         raise Exception('test2')
      6 except Exception as e:
      7     raise Exception(f'{e} 3').with_traceback(sys.exc_info()[2])

Exception: test2 3
```

</td></tr></table>

Reply:
When the traceback is longer I think it doesn't look as strange for the `with_traceback` version, but sure. `from` also seems to do some reordering, but either is easy enough to parse.

What I would really like to do, is just modify the existing error to have some extra information in the message. I don't think there's a perfect solution to this since exceptions can define their own string representation.

Reply:
yes, `type(e)(msg)` is flaky. The exception type’s `__init__` might not accept a message as its first argument.

If you understand the `with_taceback` output, please explain it to me because I don’t. The modified exception has 2 locations? huh?